File: Kotlin/GameOfLife/src/unitTests/FrameTest.kt
package unitTests

import gameOfLife.FrameMaker
import gameOfLife.entity.Frame
import gameOfLife.type.StateType
import org.junit.Test

import org.junit.Assert.*
import java.io.ByteArrayInputStream

class FrameTest {

    @Test
    fun checkNextGenForBeaconTest() {

        System.setIn(ByteArrayInputStream("i\ni\nbeacon".toByteArray()))

        val frame = FrameMaker.createFrame()

        assertEquals(8, frame.livingCount())

        assertEquals(StateType.ALIVE, frame.cells[2][2].state)
        assertEquals(StateType.ALIVE, frame.cells[3][3].state)

        frame.createNextGen()

        assertEquals(6, frame.livingCount())

        assertEquals(StateType.DEAD, frame.cells[2][2].state)
        assertEquals(StateType.DEAD, frame.cells[3][3].state)

        frame.createNextGen()

        assertEquals(8, frame.livingCount())

        assertEquals(StateType.ALIVE, frame.cells[2][2].state)
        assertEquals(StateType.ALIVE, frame.cells[3][3].state)
    }

    @Test
    fun checkNextGenForBlinkerTest() {

        System.setIn(ByteArrayInputStream("i\ni\nblinker".toByteArray()))

        val frame = FrameMaker.createFrame()

        assertEquals(3, frame.livingCount())

        assertEquals(StateType.ALIVE, frame.cells[2][1].state)
        assertEquals(StateType.ALIVE, frame.cells[2][3].state)
        assertEquals(StateType.DEAD, frame.cells[1][2].state)
        assertEquals(StateType.DEAD, frame.cells[3][2].state)

        frame.createNextGen()

        assertEquals(3, frame.livingCount())

        assertEquals(StateType.DEAD, frame.cells[2][1].state)
        assertEquals(StateType.DEAD, frame.cells[2][3].state)
        assertEquals(StateType.ALIVE, frame.cells[1][2].state)
        assertEquals(StateType.ALIVE, frame.cells[3][2].state)

        frame.createNextGen()

        assertEquals(3, frame.livingCount())

        assertEquals(StateType.ALIVE, frame.cells[2][1].state)
        assertEquals(StateType.ALIVE, frame.cells[2][3].state)
        assertEquals(StateType.DEAD, frame.cells[1][2].state)
        assertEquals(StateType.DEAD, frame.cells[3][2].state)
    }

    @Test
    fun checkNextGenForFullHouseTest() {

        System.setIn(ByteArrayInputStream("i\ni\nfullhouse".toByteArray()))

        val frame = FrameMaker.createFrame()

        assertEquals(36, frame.livingCount())

        frame.createNextGen()

        assertEquals(0, frame.livingCount())

    }

    @Test
    fun checkNextGenForToadTest() {
        System.setIn(ByteArrayInputStream("i\ni\ntoad".toByteArray()))

        val frame = FrameMaker.createFrame()

        assertEquals(6, frame.livingCount())

        assertEquals(StateType.DEAD,  frame.cells[1][3].state)

        assertEquals(StateType.DEAD,  frame.cells[2][1].state)
        assertEquals(StateType.ALIVE, frame.cells[2][2].state)
        assertEquals(StateType.ALIVE, frame.cells[2][3].state)
        assertEquals(StateType.ALIVE, frame.cells[2][4].state)

        assertEquals(StateType.ALIVE, frame.cells[3][1].state)
        assertEquals(StateType.ALIVE, frame.cells[3][2].state)
        assertEquals(StateType.ALIVE, frame.cells[3][3].state)
        assertEquals(StateType.DEAD,  frame.cells[3][4].state)

        assertEquals(StateType.DEAD,  frame.cells[4][2].state)


        frame.createNextGen()

        assertEquals(6, frame.livingCount())

        assertEquals(StateType.ALIVE, frame.cells[1][3].state)

        assertEquals(StateType.ALIVE, frame.cells[2][1].state)
        assertEquals(StateType.DEAD,  frame.cells[2][2].state)
        assertEquals(StateType.DEAD,  frame.cells[2][3].state)
        assertEquals(StateType.ALIVE, frame.cells[2][4].state)

        assertEquals(StateType.ALIVE, frame.cells[3][1].state)
        assertEquals(StateType.DEAD,  frame.cells[3][2].state)
        assertEquals(StateType.DEAD,  frame.cells[3][3].state)
        assertEquals(StateType.ALIVE, frame.cells[3][4].state)

        assertEquals(StateType.ALIVE, frame.cells[4][2].state)
    }
}

File: Lua/GameOfLife/GameOfLife/test/frameTest.lua

local old = io.read
local input = {}
local cnt = 0

io.read = function()
    cnt = cnt+1
    return input[cnt]
end

local FrameMaker = require("FrameMaker").getInstance()
local StateType = require "type.StateType"

function testCheckNextGenForBeacon()

    input = {"y", "y", "beacon", 0}
    cnt = 0

    local frame = FrameMaker.createFrame()

    lu.assertEquals(frame:getLivingCount(), 8)

    lu.assertEquals(frame.cells[3][3].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[4][4].state, StateType.ALIVE)

    frame:createNextGen()

    lu.assertEquals(frame:getLivingCount(), 6)

    lu.assertEquals(frame.cells[3][3].state, StateType.DEAD)
    lu.assertEquals(frame.cells[4][4].state, StateType.DEAD)

    frame:createNextGen()

    lu.assertEquals(frame:getLivingCount(), 8)

    lu.assertEquals(frame.cells[3][3].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[4][4].state, StateType.ALIVE)
end


function testCheckNextGenForBlinker()

    input = {"y", "y", "blinker", 0}
    cnt = 0

    local frame = FrameMaker.createFrame()

    lu.assertEquals(frame:getLivingCount(), 3)

    lu.assertEquals(frame.cells[3][2].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[3][3].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[3][4].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[2][3].state, StateType.DEAD)
    lu.assertEquals(frame.cells[4][3].state, StateType.DEAD)

    frame:createNextGen()

    lu.assertEquals(frame:getLivingCount(), 3)

    lu.assertEquals(frame.cells[3][2].state, StateType.DEAD)
    lu.assertEquals(frame.cells[3][3].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[3][4].state, StateType.DEAD)
    lu.assertEquals(frame.cells[2][3].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[4][3].state, StateType.ALIVE)

    frame:createNextGen()

    lu.assertEquals(frame:getLivingCount(), 3)

    lu.assertEquals(frame.cells[3][2].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[3][3].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[3][4].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[2][3].state, StateType.DEAD)
    lu.assertEquals(frame.cells[4][3].state, StateType.DEAD)
end


function testCheckNextGenForFullHouse()

    input = {"y", "y", "fullhouse", 0}
    cnt = 0

    local frame = FrameMaker.createFrame()

    lu.assertEquals(frame:getLivingCount(), 36)

    frame:createNextGen()

    lu.assertEquals(frame:getLivingCount(), 0)

end


function testCheckNextGenForToad()
    input = {"y", "y", "toad", 0}
    cnt = 0

    local frame = FrameMaker.createFrame()

    lu.assertEquals(frame:getLivingCount(), 6)

    lu.assertEquals(frame.cells[2][4].state, StateType.DEAD)

    lu.assertEquals(frame.cells[3][2].state, StateType.DEAD)
    lu.assertEquals(frame.cells[3][3].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[3][4].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[3][5].state, StateType.ALIVE)

    lu.assertEquals(frame.cells[4][2].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[4][3].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[4][4].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[4][5].state, StateType.DEAD)

    lu.assertEquals(frame.cells[5][3].state, StateType.DEAD)


    frame:createNextGen()

    lu.assertEquals(frame:getLivingCount(), 6)

    lu.assertEquals(frame.cells[2][4].state, StateType.ALIVE)

    lu.assertEquals(frame.cells[3][2].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[3][3].state, StateType.DEAD)
    lu.assertEquals(frame.cells[3][4].state, StateType.DEAD)
    lu.assertEquals(frame.cells[3][5].state, StateType.ALIVE)

    lu.assertEquals(frame.cells[4][2].state, StateType.ALIVE)
    lu.assertEquals(frame.cells[4][3].state, StateType.DEAD)
    lu.assertEquals(frame.cells[4][4].state, StateType.DEAD)
    lu.assertEquals(frame.cells[4][5].state, StateType.ALIVE)

    lu.assertEquals(frame.cells[5][3].state, StateType.ALIVE)
end

io.read = old
